modules/standard/DistributedDeque.chpl

```diff
@@ -0,0 +1,1000 @@
+        curr = tmp;
+      }
+
+      if cached != nil {
+        delete cached;
+      }
+    }
+  }
+}
+
+/*
+  A parallel-safe scalable distributed double-ended queue that supports both
+  insertion and removal from either end of the queue. Can be used as a Queue,
+  Stack, or even a List.
+*/
+class DistributedDeque : Collection {
+  /*
+    Capacity, the maximum number of elements a Deque can hold. A `cap` of -1 is
+    considered unbounded.
+  */
+  var cap : int;
+  pragma "no doc"
+  var targetLocDom : domain(1);
+  /*
+    Locales to distribute the `Deque` across.
+  */
+  var targetLocales : [targetLocDom] locale;
+
+  // Privatization id
+  pragma "no doc"
+  var pid : int;
+
+  // Keeps track of which slot we are on...
+  pragma "no doc"
+  var globalHead : atomic int;
+  pragma "no doc"
+  var globalTail : atomic int;
+  pragma "no doc"
+  var queueSize : atomic int;
+
+  // Freezing the queue consists of two phases: The 'marked' phase, where the queue
+  // is marked for a state change, which prevents any new tasks for a particular state
+  // from entering, followed by a 'waiting' phase where we wait any concurrent tasks
+  // to find across all nodes (in case they did not notice the state change). This
+  // applies to state changes to frozen and unfrozen state.
+  pragma "no doc"
+  var concurrentTasks : atomic int;
+  pragma "no doc"
+  var frozenState : atomic int;
+
+  // We maintain an array of slots, wherein each slot is a pointer into a node's
+  // address space. To maximize parallelism, we maintain numLocales * maxTaskPar
+  // to reduce the amount of contention.
+  pragma "no doc"
+  var nSlots : int;
+  pragma "no doc"
+  var slotSpace = {0..-1};
+  pragma "no doc"
+  var slots : [slotSpace] LocalDeque(eltType);
+
+  proc DistributedDeque(type eltType, cap : int = -1, targetLocales : [?locDom] locale =Locales) {
+    this.cap = cap;
+    this.nSlots = here.maxTaskPar * targetLocales.size;
+    this.slotSpace = {0..#this.nSlots};
+    this.targetLocDom = locDom;
+    this.targetLocales = targetLocales;
+
+    // Initialize each slot. We use a round-robin algorithm.
+    var idx : atomic int;
+    for 0 .. #here.maxTaskPar {
+      for loc in targetLocales do on loc {
+        var i = idx.fetchAdd(1);
+        slots[i] = new LocalDeque(eltType);
+      }
+    }
+
+    pid = _newPrivatizedClass(this);
+  }
+
+  pragma "no doc"
+  proc DistributedDeque(other, privData, type eltType = other.eltType) {
+    this.cap = other.cap;
+    this.targetLocDom = other.targetLocDom;
+    this.targetLocales = other.targetLocales;
+    this.nSlots = other.nSlots;
+    this.slotSpace = {0..#this.nSlots};
+    slots = other.slots;
+  }
+
+  pragma "no doc"
+  proc dsiPrivatize(privData) {
+      return new DistributedDeque(this, privData);
+  }
+
+  pragma "no doc"
+  proc dsiGetPrivatizeData() {
+    return pid;
+  }
+
+  pragma "no doc"
+  inline proc getPrivatizedThis {
+    return chpl_getPrivatizedCopy(this.type, pid);
+  }

```

Comment:
Similarly privatization methods should be marked with TODO or FIXME or something